scalapb-runtime/jvm/src/main/scala/scalapb/StructUtils.scala

```diff
@@ -0,0 +1,122 @@
+package scalapb
+
+import java.util.Base64
+
+import com.google.protobuf.ByteString
+import com.google.protobuf.struct.Value.Kind
+import com.google.protobuf.struct.{ListValue, Struct, Value}
+import scalapb.descriptors._
+
+import scala.util.{Failure, Success, Try}
+
+object StructUtils {
+  case class StructDeserError(index: Int, error: String)
+  def toStruct(generatedMessage: GeneratedMessage): Struct =
+    toStruct(generatedMessage.toPMessage.value)
+
+  def fromStruct[T <: GeneratedMessage](
+      struct: Struct
+  )(implicit companion: GeneratedMessageCompanion[T]): Either[StructDeserError, T] = {
+    val fieldDescriptorToPValue = structMapToFDMap(struct.fields)
+    fieldDescriptorToPValue.map(companion.messageReads.read)
+  }
+
+  private def structMapToFDMap(
+      structFields: Map[String, Value]
+  )(implicit companion: GeneratedMessageCompanion[_]): Either[StructDeserError, PMessage] = {
+    val fieldDescriptorToPValue = companion.scalaDescriptor.fields.map { fd =>
```

Comment:
btw, I left a few Vals in place (here and above) because I like to make debugging easier. I can inline the vals if you prefer the full fluent idiom.